tests/test_python_reducers.py

```diff
@@ -157,3 +167,205 @@ def test_reduce_with_ellipsis_can_reduce_single_block():
+
+def test_is_python_returns_true_for_valid_python():
+    """Test that is_python returns True for valid Python code."""
+    valid_code = b"x = 1"
+    assert is_python(valid_code) is True
+
+
+def test_is_python_returns_false_for_invalid_python():
+    """Test that is_python returns False for invalid Python code."""
+    invalid_code = b"def foo(:"
+    assert is_python(invalid_code) is False
+
+
+def test_libcst_transform_handles_test_case_becoming_invalid():
+    """Test that libcst_transform handles the case where the test case becomes invalid Python mid-reduction.
+
+    This exercises lines 73-75: ParserSyntaxError when parsing during can_apply.
+    The test case is valid initially but becomes invalid between calls.
+    """
+    # Track how many times current_test_case is accessed
+    access_count = [0]
+    valid_code = b"x = 1\ny = 2\nz = 3\n"
+    invalid_code = b"def foo(:\n"
+
+    class MockProblem:
+        def __init__(self):
+            self.work = WorkContext(parallelism=1)
+            self._test_case = valid_code
+
+        @property
+        def current_test_case(self):
+            access_count[0] += 1
+            # First access returns valid code (for initial parse)
+            # Later accesses return invalid code (to trigger ParserSyntaxError in can_apply)
+            if access_count[0] <= 2:
+                return valid_code
+            return invalid_code
+
+        def sort_key(self, x):
+            return (len(x), x)
+
+        async def is_interesting(self, x):
+            return True
+
+    problem = MockProblem()
+
+    async def run_test():
+        await libcst_transform(
+            problem,  # type: ignore
+            m.SimpleStatementLine(),
+            lambda x: libcst.RemoveFromParent(),
+        )
+
+    trio.run(run_test)
+    # The function should complete without raising, handling the ParserSyntaxError internally
+
+
+def test_libcst_transform_handles_cst_validation_error():
+    """Test that libcst_transform handles CSTValidationError gracefully.
+
+    This exercises line 80-81: CSTValidationError during transformation.
+    """
+    # Create a transformer that causes validation errors
+    def bad_transformer(node):
+        # Try to create an invalid CST node by adding a node in wrong context
+        # This should trigger CSTValidationError
+        raise libcst.CSTValidationError("Test validation error")
+
+    code = b"x = 1\n"
+
+    async def is_interesting(x: bytes) -> bool:
+        return True
+
+    async def run_test():
+        problem: BasicReductionProblem[bytes] = BasicReductionProblem(
+            initial=code,
+            is_interesting=is_interesting,
+            work=WorkContext(parallelism=1),
+        )
+        await libcst_transform(
+            problem,
+            m.SimpleStatementLine(),
+            bad_transformer,
+        )
+
+    trio.run(run_test)
+    # Should complete without raising
+
+
+def test_libcst_transform_handles_type_error_does_not_allow():
```

Comment:
I'm not sure why this isn't just an async test using pytest-trio. Please also look through other tests for use of trio.run to make sure there aren't more of these. Some tests do legitimately need to use trio.run, but it should be rare.